src/lib.rs

```diff
@@ -1,14 +1,180 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+mod args;
+mod context;
+mod rule;
+mod violation;
+
+use std::{
+    borrow::Cow,
+    process,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+pub use args::Args;
+use clap::Parser;
+use context::QueryMatchContext;
+use rule::{ResolvedRule, Rule, RuleBuilder, RuleListenerBuilder};
+use tree_sitter::Query;
+use violation::ViolationBuilder;
+
+use crate::context::Context;
+
+#[macro_export]
+macro_rules! regex {
+    ($re:expr $(,)?) => {{
+        static RE: std::sync::OnceLock<regex::Regex> = std::sync::OnceLock::new();
+        RE.get_or_init(|| regex::Regex::new($re).unwrap())
+    }};
+}
+
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP: &str = "_tree_sitter_lint_capture";
+const CAPTURE_NAME_FOR_TREE_SITTER_GREP_WITH_LEADING_AT: &str = "@_tree_sitter_lint_capture";
+
+pub fn run(_args: Args) {
+    let language = tree_sitter_rust::language();
+    let context = Context::new(language);
+    let resolved_rules = get_rules()
+        .into_iter()
+        .map(|rule| rule.resolve(&context))
+        .collect::<Vec<_>>();
+    let aggregated_queries = AggregatedQueries::new(&resolved_rules, &context);
+    let tree_sitter_grep_args = tree_sitter_grep::Args::parse_from([
+        "tree_sitter_grep",
+        "-q",
+        &aggregated_queries.query_text,
+        "-l",
+        "rust",
+        "--capture",
+        CAPTURE_NAME_FOR_TREE_SITTER_GREP,
```

Comment:
I'd thought we would need to make `tree-sitter-grep` be able to handle the concept of "multiple queries/captures" but nope tree-sitter queries by their nature "multiplex"/"de-multiplex" (?) (just by concatenating several top-level "patterns")

So just had to make the one change to the crate-level API to additionally expose the "pattern index" of the matching capture (which is what lets us do the "de-multiplexing")

Reply:
Ya this is definitely a good "driver" for kicking the tires on the `tree_sitter_grep` crate-level API